### DISCUSSION TITLE
Grabbing entry keyboard focus without inhibiting keyboad interactivity?

### DISCUSSION THREAD

**Discussion by ErikReider** [wmww/gtk-layer-shell#149] [Q&A] [answered]
I'm trying to add a Gtk entry to my pop up [notifications](https://github.com/ErikReider/SwayNotificationCenter) but haven't been able to grab the keyboard focus without inhibiting the keyboard interactivity.

The window that all of the notifications spawn in is an Overlay layer.

Is there any way of doing this?

**Comment by wmww** [comment, marked as answer]
> grab the keyboard focus without inhibiting the keyboard interactivity

I'm not sure exactly what this means, can you explain in more detail the behavior you do and don't want?

**Reply by ErikReider** [reply]
Oh sorry, mb. So my notifications are layered in the Overlay layer without any keyboard inhibition, so users can still interact with other software while still being able to see their notification. No issues at all.

I've now decided to add a GtkEntry to each notification to be able to have inline replies, but the GTKEntry won't grab my keyboard focus unless I set the keyboard_mode to EXCLUSIVE for the layered window. So without the keyboard inhibition, I can't type anything into said GtkEntry.

![image](https://user-images.githubusercontent.com/35975961/220450862-fa87c4e6-c072-4ba4-b270-c008c70ae661.png)

Is there any way of being able to grab the keyboard focus without keyboard inhibition? (inhibition may be the wrong word 😅)

**Reply by wmww** [reply]
Are you using `GTK_LAYER_SHELL_KEYBOARD_MODE_ON_DEMAND`? With that, are you unable to click in the text box and type in text?

**Reply by ErikReider** [reply]
Never really understood what it does due to it behaving the same as EXCLUSIVE on Sway. So I'm guessing that this is a Sway limitation at the moment?

**Reply by wmww** [reply]
Ah right, see https://github.com/wmww/gtk-layer-shell/issues/141#issuecomment-1341400400.

One thing that could *maybe* work is to disable keyboard interactivity, and when the text box is clicked turn it to exclusive and also create a popup. When the popup is dismissed it means the user clicked outside of your surface and you should set keyboard interactivity back to disabled. This would be janky, but something like that *might* allow you to approximate the behavior you want.

The real solution is to poke the Sway devs to support layer shell v4

**Reply by ErikReider** [reply]
Interesting! I'll look into it! Thanks for your time :)

**Reply by Tamaranch** [reply]
> The real solution is to poke the Sway devs to support layer shell v4

See also https://github.com/swaywm/sway/pull/5974

**Comment by ErikReider** [comment]
For anyone else experiencing this, [Sway PR](https://github.com/swaywm/sway/pull/7587) adds on demand support which would fix this issue. So until it's merged, please help testing to hopefully get it through the gates :)
